File: pocs/auth0-kotlin/src/main/kotlin/com/auth0/kotlin/Auth0.kt
package com.auth0.kotlin

import com.auth0.client.auth.AuthAPI
import com.auth0.client.auth.AuthorizeUrlBuilder
import com.auth0.client.auth.ClientAssertionSigner
import com.auth0.client.auth.LogoutUrlBuilder
import com.auth0.client.mgmt.ActionsClient
import com.auth0.client.mgmt.BrandingClient
import com.auth0.client.mgmt.ClientGrantsClient
import com.auth0.client.mgmt.ClientsClient
import com.auth0.client.mgmt.ConnectionProfilesClient
import com.auth0.client.mgmt.ConnectionsClient
import com.auth0.client.mgmt.CustomDomainsClient
import com.auth0.client.mgmt.DeviceCredentialsClient
import com.auth0.client.mgmt.EmailTemplatesClient
import com.auth0.client.mgmt.EventStreamsClient
import com.auth0.client.mgmt.EventsClient
import com.auth0.client.mgmt.FlowsClient
import com.auth0.client.mgmt.FormsClient
import com.auth0.client.mgmt.GroupsClient
import com.auth0.client.mgmt.HooksClient
import com.auth0.client.mgmt.JobsClient
import com.auth0.client.mgmt.LogStreamsClient
import com.auth0.client.mgmt.LogsClient
import com.auth0.client.mgmt.ManagementApi
import com.auth0.client.mgmt.ManagementApiBuilder
import com.auth0.client.mgmt.NetworkAclsClient
import com.auth0.client.mgmt.OrganizationsClient
import com.auth0.client.mgmt.PromptsClient
import com.auth0.client.mgmt.RateLimitPoliciesClient
import com.auth0.client.mgmt.RefreshTokensClient
import com.auth0.client.mgmt.ResourceServersClient
import com.auth0.client.mgmt.RolesClient
import com.auth0.client.mgmt.RulesClient
import com.auth0.client.mgmt.RulesConfigsClient
import com.auth0.client.mgmt.SelfServiceProfilesClient
import com.auth0.client.mgmt.SessionsClient
import com.auth0.client.mgmt.StatsClient
import com.auth0.client.mgmt.SupplementalSignalsClient
import com.auth0.client.mgmt.TicketsClient
import com.auth0.client.mgmt.TokenExchangeProfilesClient
import com.auth0.client.mgmt.UserAttributeProfilesClient
import com.auth0.client.mgmt.UserBlocksClient
import com.auth0.client.mgmt.UserGrantsClient
import com.auth0.client.mgmt.UsersClient
import com.auth0.client.mgmt.anomaly.AnomalyClient
import com.auth0.client.mgmt.attackprotection.AttackProtectionClient
import com.auth0.client.mgmt.core.Environment
import com.auth0.client.mgmt.core.LogConfig
import com.auth0.client.mgmt.emails.EmailsClient
import com.auth0.client.mgmt.guardian.GuardianClient
import com.auth0.client.mgmt.keys.KeysClient
import com.auth0.client.mgmt.riskassessments.RiskAssessmentsClient
import com.auth0.client.mgmt.tenants.TenantsClient
import com.auth0.client.mgmt.verifiablecredentials.VerifiableCredentialsClient
import com.auth0.net.Request
import com.auth0.net.Response
import com.auth0.net.client.Auth0HttpClient
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.OkHttpClient

@DslMarker
public annotation class Auth0Dsl

public object Auth0 {
    public fun auth(
        domain: String,
        clientId: String,
        configure: AuthenticationConfig.() -> Unit = {},
    ): AuthClient =
        AuthAPI.newBuilder(domain, clientId).configured(configure)

    public fun auth(
        domain: String,
        clientId: String,
        clientSecret: String,
        configure: AuthenticationConfig.() -> Unit = {},
    ): AuthClient =
        AuthAPI.newBuilder(domain, clientId, clientSecret).configured(configure)

    public fun auth(
        domain: String,
        clientId: String,
        signer: ClientAssertionSigner,
        configure: AuthenticationConfig.() -> Unit = {},
    ): AuthClient =
        AuthAPI.newBuilder(domain, clientId, signer).configured(configure)

    public fun managementWithToken(
        domain: String,
        token: String,
        configure: ManagementConfig.() -> Unit = {},
    ): ManagementClient =
        ManagementApi.builder().domain(domain).token(token).configured(configure)

    public fun managementWithClientCredentials(
        domain: String,
        clientId: String,
        clientSecret: String,
        configure: ManagementConfig.() -> Unit = {},
    ): ManagementClient =
        ManagementApi.builder().domain(domain).clientCredentials(clientId, clientSecret).configured(configure)

    public fun management(configure: ManagementConfig.() -> Unit): ManagementClient =
        ManagementApi.builder().configured(configure)
}

@Auth0Dsl
public class AuthenticationConfig internal constructor(private val builder: AuthAPI.Builder) {
    public fun httpClient(client: Auth0HttpClient): Unit {
        builder.withHttpClient(client)
    }
}

@Auth0Dsl
public class ManagementConfig internal constructor(private val builder: ManagementApiBuilder) {
    public fun domain(value: String): Unit {
        builder.domain(value)
    }

    public fun url(value: String): Unit {
        builder.url(value)
    }

    public fun environment(value: Environment): Unit {
        builder.environment(value)
    }

    public fun token(value: String): Unit {
        builder.token(value)
    }

    public fun clientCredentials(clientId: String, clientSecret: String): Unit {
        builder.clientCredentials(clientId, clientSecret)
    }

    public fun audience(value: String): Unit {
        builder.audience(value)
    }

    public fun timeout(seconds: Int): Unit {
        builder.timeout(seconds)
    }

    public fun maxRetries(value: Int): Unit {
        builder.maxRetries(value)
    }

    public fun customDomain(value: String): Unit {
        builder.customDomain(value)
    }

    public fun httpClient(client: OkHttpClient): Unit {
        builder.httpClient(client)
    }

    public fun logging(config: LogConfig): Unit {
        builder.logging(config)
    }

    public fun header(name: String, value: String): Unit {
        builder.addHeader(name, value)
    }
}

public class AuthClient(public val java: AuthAPI) {
    public fun authorizationUrl(
        redirectUri: String,
        configure: AuthorizeUrlBuilder.() -> Unit = {},
    ): String =
        java.authorizeUrl(redirectUri).apply(configure).build()

    public fun logoutUrl(
        returnToUrl: String,
        includeClientId: Boolean = true,
        configure: LogoutUrlBuilder.() -> Unit = {},
    ): String =
        java.logoutUrl(returnToUrl, includeClientId).apply(configure).build()

    public inline fun <T> use(block: AuthAPI.() -> T): T =
        java.block()
}

public class ManagementClient(public val java: ManagementApi) {
    public val actions: ActionsClient get() = java.actions()
    public val branding: BrandingClient get() = java.branding()
    public val clientGrants: ClientGrantsClient get() = java.clientGrants()
    public val clients: ClientsClient get() = java.clients()
    public val connectionProfiles: ConnectionProfilesClient get() = java.connectionProfiles()
    public val connections: ConnectionsClient get() = java.connections()
    public val customDomains: CustomDomainsClient get() = java.customDomains()
    public val deviceCredentials: DeviceCredentialsClient get() = java.deviceCredentials()
    public val emailTemplates: EmailTemplatesClient get() = java.emailTemplates()
    public val eventStreams: EventStreamsClient get() = java.eventStreams()
    public val events: EventsClient get() = java.events()
    public val flows: FlowsClient get() = java.flows()
    public val forms: FormsClient get() = java.forms()
    public val userGrants: UserGrantsClient get() = java.userGrants()
    public val groups: GroupsClient get() = java.groups()
    public val hooks: HooksClient get() = java.hooks()
    public val jobs: JobsClient get() = java.jobs()
    public val logStreams: LogStreamsClient get() = java.logStreams()
    public val logs: LogsClient get() = java.logs()
    public val networkAcls: NetworkAclsClient get() = java.networkAcls()
    public val organizations: OrganizationsClient get() = java.organizations()
    public val prompts: PromptsClient get() = java.prompts()
    public val rateLimitPolicies: RateLimitPoliciesClient get() = java.rateLimitPolicies()
    public val refreshTokens: RefreshTokensClient get() = java.refreshTokens()
    public val resourceServers: ResourceServersClient get() = java.resourceServers()
    public val roles: RolesClient get() = java.roles()
    public val rules: RulesClient get() = java.rules()
    public val rulesConfigs: RulesConfigsClient get() = java.rulesConfigs()
    public val selfServiceProfiles: SelfServiceProfilesClient get() = java.selfServiceProfiles()
    public val sessions: SessionsClient get() = java.sessions()
    public val stats: StatsClient get() = java.stats()
    public val supplementalSignals: SupplementalSignalsClient get() = java.supplementalSignals()
    public val tickets: TicketsClient get() = java.tickets()
    public val tokenExchangeProfiles: TokenExchangeProfilesClient get() = java.tokenExchangeProfiles()
    public val userAttributeProfiles: UserAttributeProfilesClient get() = java.userAttributeProfiles()
    public val userBlocks: UserBlocksClient get() = java.userBlocks()
    public val users: UsersClient get() = java.users()
    public val anomaly: AnomalyClient get() = java.anomaly()
    public val attackProtection: AttackProtectionClient get() = java.attackProtection()
    public val emails: EmailsClient get() = java.emails()
    public val guardian: GuardianClient get() = java.guardian()
    public val keys: KeysClient get() = java.keys()
    public val riskAssessments: RiskAssessmentsClient get() = java.riskAssessments()
    public val tenants: TenantsClient get() = java.tenants()
    public val verifiableCredentials: VerifiableCredentialsClient get() = java.verifiableCredentials()

    public inline fun <T> use(block: ManagementApi.() -> T): T =
        java.block()
}

public fun AuthAPI.asKotlin(): AuthClient =
    AuthClient(this)

public fun ManagementApi.asKotlin(): ManagementClient =
    ManagementClient(this)

public fun <T> Request<T>.executeBody(): T? =
    execute().body

public suspend fun <T> Request<T>.await(): Response<T> =
    suspendCancellableCoroutine { continuation ->
        val future = executeAsync()
        continuation.invokeOnCancellation { future.cancel(true) }
        future.whenComplete { response, error ->
            if (error == null) {
                continuation.resume(response)
            } else {
                continuation.resumeWithException(error.cause ?: error)
            }
        }
    }

public suspend fun <T> Request<T>.awaitBody(): T? =
    await().body

private fun AuthAPI.Builder.configured(configure: AuthenticationConfig.() -> Unit): AuthClient {
    AuthenticationConfig(this).configure()
    return AuthClient(build())
}

private fun ManagementApiBuilder.configured(configure: ManagementConfig.() -> Unit): ManagementClient {
    ManagementConfig(this).configure()
    return ManagementClient(build())
}

File: pocs/auth0-scala3/src/main/scala/auth0scala3/Auth0.scala
package auth0scala3

import com.auth0.client.auth.AuthAPI
import com.auth0.client.auth.AuthorizeUrlBuilder
import com.auth0.client.auth.ClientAssertionSigner
import com.auth0.client.auth.LogoutUrlBuilder
import com.auth0.client.mgmt.ManagementApi
import com.auth0.client.mgmt.ManagementApiBuilder
import com.auth0.client.mgmt.core.Environment
import com.auth0.client.mgmt.core.LogConfig
import com.auth0.net.Request
import com.auth0.net.Response
import com.auth0.net.client.Auth0HttpClient
import okhttp3.OkHttpClient
import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.jdk.FutureConverters.*

object Auth0:
  def auth(domain: String, clientId: String): AuthClient =
    auth(domain, clientId, _ => ())

  def auth(
      domain: String,
      clientId: String,
      configure: AuthenticationConfig => Unit
  ): AuthClient =
    configured(AuthAPI.newBuilder(domain, clientId), configure)

  def auth(domain: String, clientId: String, clientSecret: String): AuthClient =
    auth(domain, clientId, clientSecret, _ => ())

  def auth(
      domain: String,
      clientId: String,
      clientSecret: String,
      configure: AuthenticationConfig => Unit
  ): AuthClient =
    configured(AuthAPI.newBuilder(domain, clientId, clientSecret), configure)

  def auth(domain: String, clientId: String, signer: ClientAssertionSigner): AuthClient =
    auth(domain, clientId, signer, _ => ())

  def auth(
      domain: String,
      clientId: String,
      signer: ClientAssertionSigner,
      configure: AuthenticationConfig => Unit
  ): AuthClient =
    configured(AuthAPI.newBuilder(domain, clientId, signer), configure)

  def managementWithToken(
      domain: String,
      token: String,
      configure: ManagementConfig => Unit = _ => ()
  ): ManagementClient =
    configured(ManagementApi.builder().domain(domain).token(token), configure)

  def managementWithClientCredentials(
      domain: String,
      clientId: String,
      clientSecret: String,
      configure: ManagementConfig => Unit = _ => ()
  ): ManagementClient =
    configured(ManagementApi.builder().domain(domain).clientCredentials(clientId, clientSecret), configure)

  def management(configure: ManagementConfig => Unit): ManagementClient =
    configured(ManagementApi.builder(), configure)

  private def configured(
      builder: AuthAPI.Builder,
      configure: AuthenticationConfig => Unit
  ): AuthClient =
    configure(AuthenticationConfig(builder))
    AuthClient(builder.build())

  private def configured(
      builder: ManagementApiBuilder,
      configure: ManagementConfig => Unit
  ): ManagementClient =
    configure(ManagementConfig(builder))
    ManagementClient(builder.build())

final class AuthenticationConfig private (builder: AuthAPI.Builder):
  def httpClient(client: Auth0HttpClient): Unit =
    builder.withHttpClient(client)

object AuthenticationConfig:
  private[auth0scala3] def apply(builder: AuthAPI.Builder): AuthenticationConfig =
    new AuthenticationConfig(builder)

final class ManagementConfig private (builder: ManagementApiBuilder):
  def domain(value: String): Unit =
    builder.domain(value)

  def url(value: String): Unit =
    builder.url(value)

  def environment(value: Environment): Unit =
    builder.environment(value)

  def token(value: String): Unit =
    builder.token(value)

  def clientCredentials(clientId: String, clientSecret: String): Unit =
    builder.clientCredentials(clientId, clientSecret)

  def audience(value: String): Unit =
    builder.audience(value)

  def timeout(seconds: Int): Unit =
    builder.timeout(seconds)

  def maxRetries(value: Int): Unit =
    builder.maxRetries(value)

  def customDomain(value: String): Unit =
    builder.customDomain(value)

  def httpClient(client: OkHttpClient): Unit =
    builder.httpClient(client)

  def logging(config: LogConfig): Unit =
    builder.logging(config)

  def header(name: String, value: String): Unit =
    builder.addHeader(name, value)

object ManagementConfig:
  private[auth0scala3] def apply(builder: ManagementApiBuilder): ManagementConfig =
    new ManagementConfig(builder)

final class AuthorizationConfig private (builder: AuthorizeUrlBuilder):
  def connection(value: String): Unit =
    builder.withConnection(value)

  def audience(value: String): Unit =
    builder.withAudience(value)

  def state(value: String): Unit =
    builder.withState(value)

  def scope(value: String): Unit =
    builder.withScope(value)

  def responseType(value: String): Unit =
    builder.withResponseType(value)

  def organization(value: String): Unit =
    builder.withOrganization(value)

  def invitation(value: String): Unit =
    builder.withInvitation(value)

  def parameter(name: String, value: String): Unit =
    builder.withParameter(name, value)

  def codeChallenge(value: String): Unit =
    builder.withCodeChallenge(value)

object AuthorizationConfig:
  private[auth0scala3] def apply(builder: AuthorizeUrlBuilder): AuthorizationConfig =
    new AuthorizationConfig(builder)

final class LogoutConfig private (builder: LogoutUrlBuilder):
  def federated(value: Boolean = true): Unit =
    builder.useFederated(value)

object LogoutConfig:
  private[auth0scala3] def apply(builder: LogoutUrlBuilder): LogoutConfig =
    new LogoutConfig(builder)

final class AuthClient(val java: AuthAPI):
  def authorizationUrl(
      redirectUri: String,
      configure: AuthorizationConfig => Unit = _ => ()
  ): String =
    val builder = java.authorizeUrl(redirectUri)
    configure(AuthorizationConfig(builder))
    builder.build()

  def logoutUrl(
      returnToUrl: String,
      includeClientId: Boolean = true,
      configure: LogoutConfig => Unit = _ => ()
  ): String =
    val builder = java.logoutUrl(returnToUrl, includeClientId)
    configure(LogoutConfig(builder))
    builder.build()

  def use[A](operation: AuthAPI => A): A =
    operation(java)

object AuthClient:
  def apply(java: AuthAPI): AuthClient =
    new AuthClient(java)

final class ManagementClient(val java: ManagementApi):
  def actions = java.actions()
  def branding = java.branding()
  def clientGrants = java.clientGrants()
  def clients = java.clients()
  def connectionProfiles = java.connectionProfiles()
  def connections = java.connections()
  def customDomains = java.customDomains()
  def deviceCredentials = java.deviceCredentials()
  def emailTemplates = java.emailTemplates()
  def eventStreams = java.eventStreams()
  def events = java.events()
  def flows = java.flows()
  def forms = java.forms()
  def userGrants = java.userGrants()
  def groups = java.groups()
  def hooks = java.hooks()
  def jobs = java.jobs()
  def logStreams = java.logStreams()
  def logs = java.logs()
  def networkAcls = java.networkAcls()
  def organizations = java.organizations()
  def prompts = java.prompts()
  def rateLimitPolicies = java.rateLimitPolicies()
  def refreshTokens = java.refreshTokens()
  def resourceServers = java.resourceServers()
  def roles = java.roles()
  def rules = java.rules()
  def rulesConfigs = java.rulesConfigs()
  def selfServiceProfiles = java.selfServiceProfiles()
  def sessions = java.sessions()
  def stats = java.stats()
  def supplementalSignals = java.supplementalSignals()
  def tickets = java.tickets()
  def tokenExchangeProfiles = java.tokenExchangeProfiles()
  def userAttributeProfiles = java.userAttributeProfiles()
  def userBlocks = java.userBlocks()
  def users = java.users()
  def anomaly = java.anomaly()
  def attackProtection = java.attackProtection()
  def emails = java.emails()
  def guardian = java.guardian()
  def keys = java.keys()
  def riskAssessments = java.riskAssessments()
  def tenants = java.tenants()
  def verifiableCredentials = java.verifiableCredentials()

  def use[A](operation: ManagementApi => A): A =
    operation(java)

object ManagementClient:
  def apply(java: ManagementApi): ManagementClient =
    new ManagementClient(java)

extension (java: AuthAPI)
  def asScala: AuthClient =
    AuthClient(java)

extension (java: ManagementApi)
  def asScala: ManagementClient =
    ManagementClient(java)

extension [A](request: Request[A])
  def executeBody(): A =
    request.execute().getBody

  def future: Future[Response[A]] =
    request.executeAsync().asScala

  def bodyFuture(using ExecutionContext): Future[A] =
    future.map(_.getBody)

extension [A](response: Response[A])
  def statusCode: Int =
    response.getStatusCode

  def bodyOption: Option[A] =
    Option(response.getBody)
